common/src/main/java/io/netty/util/internal/ConstantTimeUtils.java

```diff
@@ -0,0 +1,131 @@
+ * under the License.
+ */
+package io.netty.util.internal;
+
+public final class ConstantTimeUtils {
+    private ConstantTimeUtils() { }
+
+    /**
+     * Compare two {@code int}s without leaking timing information.
+     * <p>
+     * The {@code int} return type is intentional and is designed to allow cascading of constant time operations:
+     * <pre>
+     *     int v1 = 1;
+     *     int v1 = 1;
+     *     int v1 = 1;
+     *     int v1 = 500;
+     *     boolean equals = (equalsConstantTime(l1, l2) & equalsConstantTime(l3, l4)) != 0;
+     * </pre>
+     * @param x the first value.
+     * @param y the second value.
+     * @return {@code 0} if not equal. {@code 1} if equal.
+     */
+    public static int equalsConstantTime(int x, int y) {
+        int z = -1 ^ (x ^ y);
+        z &= z >> 16;
+        z &= z >> 8;
+        z &= z >> 4;
+        z &= z >> 2;
+        z &= z >> 1;
+        return z & 1;
+    }
+
+    /**
+     * Compare two {@code longs}s without leaking timing information.
+     * <p>
+     * The {@code int} return type is intentional and is designed to allow cascading of constant time operations:
+     * <pre>
+     *     long v1 = 1;
+     *     long v1 = 1;
+     *     long v1 = 1;
+     *     long v1 = 500;
+     *     boolean equals = (equalsConstantTime(l1, l2) & equalsConstantTime(l3, l4)) != 0;
+     * </pre>
+     * @param x the first value.
+     * @param y the second value.
+     * @return {@code 0} if not equal. {@code 1} if equal.
+     */
+    public static int equalsConstantTime(long x, long y) {
+        long z = -1L ^ (x ^ y);
+        z &= z >> 32;
+        z &= z >> 16;
+        z &= z >> 8;
+        z &= z >> 4;
+        z &= z >> 2;
+        z &= z >> 1;
+        return (int) (z & 1);
+    }
+
+    /**
+     * Compare two {@code byte} arrays for equality without leaking timing information.
+     * For performance reasons no bounds checking on the parameters is performed.
+     * <p>
+     * The {@code int} return type is intentional and is designed to allow cascading of constant time operations:
+     * <pre>
+     *     byte[] s1 = new {1, 2, 3};
+     *     byte[] s2 = new {1, 2, 3};
+     *     byte[] s3 = new {1, 2, 3};
+     *     byte[] s4 = new {4, 5, 6};
+     *     boolean equals = (equalsConstantTime(s1, 0, s2, 0, s1.length) &
+     *                       equalsConstantTime(s3, 0, s4, 0, s3.length)) != 0;
+     * </pre>
+     * @param bytes1 the first byte array.
+     * @param startPos1 the position (inclusive) to start comparing in {@code bytes1}.
+     * @param bytes2 the second byte array.
+     * @param startPos2 the position (inclusive) to start comparing in {@code bytes2}.
+     * @param length the amount of bytes to compare. This is assumed to be validated as not going out of bounds
+     * by the caller.
+     * @return {@code 0} if not equal. {@code 1} if equal.
+     */
+    public static int equalsConstantTime(byte[] bytes1, int startPos1,
```

Comment:
The name constant implies that the contents of the array does not affect the time it takes to process.  There are always going to be compiler tricks or processor tricks or OS tricks that will mess with the timing.  The important thing is to not have different behavior based on the contents.  That's where the "constant" comes from.

That said, to @buchgr's comment, I think you may still need to do a bounds check.   Something like bytes1.length - startPos1 == bytes2.length - start2Pos.